_posts/2020-10-05-how-to-write-a-new-foreman-ansible-module-in-20-lines-of-code.md

```diff
@@ -0,0 +1,161 @@
+
+* `name`, required for create, optional for update
+* `url`, required for create, optional for update
+* `username`, optional
+* `password`, optional
+* `location_ids`, optional
+* `organization_ids`, optional
+
+We also see that there are no other actions besides list, show, create, update, and delete.
+
+## Creating a minimal module
+
+A regular Ansible module always starts with an instance of `AnsibleModule`.
+
+In Foreman Ansible Modules, we have several classes based on this basic one, so we need to decide which base class from [`foreman_helper`](https://github.com/theforeman/foreman-ansible-modules/blob/develop/plugins/module_utils/foreman_helper.py) we should take.
+
+HTTP Proxies inside of Foreman are stateful (can be created and deleted) entities that can have taxonomies (Organizations and Locations), so we're taking the [`ForemanTaxonomicEntityAnsibleModule`](https://theforeman.github.io/foreman-ansible-modules/develop/api.html#foreman_helper.ForemanTaxonomicEntityAnsibleModule) class, which will provide us with the `state`, `organizations`, and `locations` parameters and automatic handling of those. Please refer to the [`foreman_helper` documentation](https://theforeman.github.io/foreman-ansible-modules/develop/api.html) for other possible base classes and their use-cases.
+
+```python
+from ansible_collections.theforeman.foreman.plugins.module_utils.foreman_helper import ForemanTaxonomicEntityAnsibleModule
+```
+
+As the helper decides which API endpoint to talk to based on the class name, we create a minimal class:
+
+```python
+class ForemanHttpProxyModule(ForemanTaxonomicEntityAnsibleModule):
+    pass
+```
+
+Now we need to add the other fields the API supports as parameters. We'll mark the `name` parameter as `required` because that's the one we'll be using for finding existing proxies -- the API has it as `optional` because it uses the ID of the proxy as the primary identification, but nobody wants to write static ids in Ansible playbooks. The `url` parameter is only required for *new* proxies, so we'll mark it as `optional` in the module parameters and document the requirement for creation. As our modules already have `username` and `password` parameters for authenticating against Foreman, we'll have to name the parameters for those fields differently -- `proxy_` seems like a good prefix to add here. And last but not least, `password` parameters should be marked as `no_log=True` so that Ansible skips them when printing debugging information about the module invocation:
+
+```python
+    module = ForemanHttpProxyModule(
+        foreman_spec=dict(
+            name=dict(required=True),
+            url=dict(),
+            proxy_username=dict(flat_name='username'),
+            proxy_password=dict(no_log=True, flat_name='password'),
+        ),
+    )
+```
+
+`flat_name` is telling our abstraction layer that the field name in the API differs from the Ansible parameter name.
```

Comment:
Again a reason to mention the documentation? There are a lot of additional options. Also we support everything argument_spec does. Maybe we even state that foremen_spec is a replacement.